In [2]:
import pandas as pd

dataset 불러오기

In [3]:
dataset1=pd.read_excel('상관분석_매출통합 (1).xlsx',index_col=0)
dataset1

session s_zone  s_zone_code      s_zone_name   total_sale  t_household  \
year                                                                           
2021        4   골목상권      2111090          강일동주민센터   6164144274          826   
2021        4   골목상권      2111089     상일여고(상일초등학교)  18215725779         1707   
2021        4   골목상권      2111088            고덕동성당   3130903256          967   
2021        4   골목상권      2111087  고덕중학교(고덕2동주민센터)   2770689276         1224   
2021        4   골목상권      2111086             한영중고   4831084227         1056   
...       ...    ...          ...              ...          ...          ...   
2017        1   골목상권      2110005              세검정   2528872295          759   
2017        1   골목상권      2110004           대신고등학교    925316009          258   
2017        1   골목상권      2110003          세검정초등학교   2420562470         1273   
2017        1   골목상권      2110002          독립문역 1번   8522048012          349   
2017        1   골목상권      2110001           이북5도청사   1803437452          436   

      income_range     t_spent  office  bank  ...  apt_0.1b_under  apt_0.1b  \
year                                          ...                             
2021             6           0       4     1  ...               0         0   
2021             6  1310650750       1     0  ...              64       402   
2021             5   941528487       0     0  ...               6        76   
2021             6   995226881       1     0  ...               1       125   
2021             6  1264810182       0     0  ...              11       136   
...            ...         ...     ...   ...  ...             ...       ...   
2017             7  1713487382       2     1  ...              91       180   
2017             6   604905596       0     0  ...              12        16   
2017             7  2635767850       0     0  ...              70       374   
2017             7   516760357       2     0  ...               8        36   
2017             8  1079739057       0     0  ...              12        47   

      apt_0.2b  apt_0.3b  apt_0.4b  apt_0.5b  apt_0.6b  apt_avg_width  \
year                                                                    
2021         0         0         0         0         0              0   
2021        56         0         0         0         0             47   
2021        30         0         0         0         0             46   
2021         9         2         1         0         0             45   
2021        49         2         0         0         0             43   
...        ...       ...       ...       ...       ...            ...   
2017        64        46        12        17        46             92   
2017         0         0         0         0         0             50   
2017       107        42        50        18         1             81   
2017        20         0        27        22         0             70   
2017        26       144        26         9        40            122   

      apt_avg_price  total_living  
year                               
2021              0         70137  
2021      152180841        861179  
2021      176342376        435291  
2021      180259184        192202  
2021      171494591        305276  
...             ...           ...  
2017      242615714        143249  
2017      110733333        353902  
2017      194931818        374027  
2017      208056122        268151  
2017      371391608        148356  

[21772 rows x 43 columns]

test data 와 train data 나누기
train data = 17년도 1분기 ~ 21년도 2분기
val data = 21년도 3분기
test data = 21년도 4분기

xdata= 전체 컬럼에서 session, s_zone, s_zone_code,s_zone_name, total_sale을 뺀 data,
ydata = totla_sale data  

In [4]:
X_dataset=dataset1.drop(columns=['session','total_sale','s_zone','s_zone_code','s_zone_name'])
y_dataset=dataset1['total_sale']

In [5]:
X_test=X_dataset[:1086]
X_train=X_dataset[2174:]
X_val=X_dataset[1086:2174]
y_val=y_dataset[1086:2174]
y_test=y_dataset[:1086]
y_train=y_dataset[2174:]


In [12]:
print("X_train",X_train.shape)
print("y_train",y_train.shape)
print("X_val",X_val.shape)
print("y_val",y_val.shape)
print("X_test",X_test.shape)
print("y_test",y_test.shape)

X_train (19598, 38)
y_train (19598,)
X_val (1088, 38)
y_val (1088,)
X_test (1086, 38)
y_test (1086,)


linear regression model with raw data

In [13]:
from sklearn.linear_model import LinearRegression
lr = LinearRegression().fit(X_train, y_train)

In [14]:
y_train_hat = lr.predict(X_train)
print(y_train)
print(y_train_hat)

year
2021     6136905604
2021    21934115430
2021     9885562997
2021     2893691821
2021     4654682074
           ...     
2017     2528872295
2017      925316009
2017     2420562470
2017     8522048012
2017     1803437452
Name: total_sale, Length: 19598, dtype: int64
[7.70896007e+09 5.53777310e+09 1.71273025e+09 ... 5.57186816e+09
 7.08910184e+09 2.37940539e+09]


In [15]:
y_val_hat = lr.predict(X_val)
print(y_val)
print(y_val_hat)

year
2021     5961650254
2021    17049532868
2021     9672498575
2021     2928283066
2021     4551502196
           ...     
2021     1876881076
2021      744236445
2021     1730866844
2021     7979637934
2021     1529373991
Name: total_sale, Length: 1088, dtype: int64
[7.70702111e+09 5.36148297e+09 1.63463805e+09 ... 6.77922430e+09
 5.88918274e+09 2.77786152e+09]


In [16]:
from sklearn.metrics import mean_absolute_error, mean_squared_error, r2_score

print('performance for TRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_train_hat))
print('train RMSE : ', mean_squared_error(y_train, y_train_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_train_hat))

print('performance for VAL--------')
print('val MAE : ', mean_absolute_error(y_val, y_val_hat))
print('val RMSE : ', mean_squared_error(y_val, y_val_hat)**0.5)
print('val R_square : ', r2_score(y_val, y_val_hat))

performance for TRAIN--------
train MAE :  2191918759.671748
train RMSE :  3554464401.714925
train R_square :  0.5614943026057684
performance for VAL--------
val MAE :  2279884116.456788
val RMSE :  3888471162.715774
val R_square :  0.4730787755886081


모델성능 파악 rmse : 3888471162.715774, val r_square:0.4730787755886081

다중공산성 판단을 위해 vif값 구하기

In [17]:
from statsmodels.stats.outliers_influence import variance_inflation_factor

vif = pd.DataFrame()
vif["VIF Factor"] = [variance_inflation_factor(X_train.values,i) for i in range(X_train.shape[1])]
vif["features"]= X_train.columns

c:\Anaconda\lib\site-packages\statsmodels\stats\outliers_influence.py:193: RuntimeWarning: divide by zero encountered in double_scalars
  vif = 1. / (1. - r_squared_i)


In [18]:
vif.round(1)

VIF Factor                features
0         20.8             t_household
1         16.3            income_range
2         -0.0                 t_spent
3          1.9                  office
4          1.5                    bank
5          2.9                 medical
6          1.3                     edu
7          1.2                   store
8          4.0                 traffic
9          4.8                office_b
10         6.4                  bank_b
11        10.0               medical_b
12         1.8                   edu_b
13         1.7                 store_b
14        11.5               traffic_b
15        21.6           t_household_b
16         1.8              work_10num
17         2.7    work_20_30_40_50_num
18         2.0             work_60_num
19         1.9            work_10num_b
20         4.2  work_20_30_40_50_num_b
21         5.6           work_60_num_b
22        12.9       apt_complexes_num
23         inf            apt_66_under
24         inf                  apt_66
25         inf                  apt_99
26         inf                 apt_132
27         inf                 apt_165
28         inf          apt_0.1b_under
29         inf                apt_0.1b
30         inf                apt_0.2b
31         inf                apt_0.3b
32         inf                apt_0.4b
33         inf                apt_0.5b
34         inf                apt_0.6b
35        22.6           apt_avg_width
36         0.1           apt_avg_price
37        10.2            total_living

vif 값이 10이상인 컬럼(t_household, income_range,medical_b, traffic_b,t_household_b,apt_complexes_num,apt_66_under, apt_66,99,132, apt_0.1b_under apt_0.1b,0.2b,0.3b,0.4b,0.5b,0.6b, apt_avg_width, total_living)을 제거하여 rmse, r_squared 변화 비교하기 

In [19]:
vifX = vif.iloc[vif['VIF Factor'].nlargest(20).index].features.values.tolist()
vifX

['apt_66_under',
 'apt_66',
 'apt_99',
 'apt_132',
 'apt_165',
 'apt_0.1b_under',
 'apt_0.1b',
 'apt_0.2b',
 'apt_0.3b',
 'apt_0.4b',
 'apt_0.5b',
 'apt_0.6b',
 'apt_avg_width',
 't_household_b',
 't_household',
 'income_range',
 'apt_complexes_num',
 'traffic_b',
 'total_living',
 'medical_b']

In [20]:
X_vif_train=X_train.drop(columns=vifX)
X_vif_val=X_val.drop(columns=vifX)

In [21]:
lr = LinearRegression().fit(X_vif_train, y_train)
y_vif_train_hat = lr.predict(X_vif_train)
print(y_train)
print(y_vif_train_hat)
y_vif_val_hat = lr.predict(X_vif_val)
print(y_val)
print(y_vif_val_hat)
print('performance for vifTRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_vif_train_hat))
print('train RMSE : ', mean_squared_error(y_train, y_vif_train_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_vif_train_hat))

print('performance for vifVAL--------')
print('val MAE : ', mean_absolute_error(y_val, y_vif_val_hat))
print('val RMSE : ', mean_squared_error(y_val, y_vif_val_hat)**0.5)
print('val R_square : ', r2_score(y_val, y_vif_val_hat))


year
2021     6136905604
2021    21934115430
2021     9885562997
2021     2893691821
2021     4654682074
           ...     
2017     2528872295
2017      925316009
2017     2420562470
2017     8522048012
2017     1803437452
Name: total_sale, Length: 19598, dtype: int64
[8.10863988e+09 6.00412534e+09 1.84904063e+09 ... 5.53376967e+09
 7.44988864e+09 2.36515088e+09]
year
2021     5961650254
2021    17049532868
2021     9672498575
2021     2928283066
2021     4551502196
           ...     
2021     1876881076
2021      744236445
2021     1730866844
2021     7979637934
2021     1529373991
Name: total_sale, Length: 1088, dtype: int64
[8.10863988e+09 6.00412534e+09 1.84904063e+09 ... 6.42277314e+09
 6.06094000e+09 2.80130082e+09]
performance for vifTRAIN--------
train MAE :  2293798198.8432045
train RMSE :  3711227351.035447
train R_square :  0.5219624382195396
performance for vifVAL--------
val MAE :  2346696084.4283543
val RMSE :  4017940740.1925507
val R_square :  0.43740614847863246


vif 10이상인 변수 제거후 모델 성능 파악 : rmse:4017940740.1925507, r_square:0.43740614847863246

vif가 10이상인 데이터 즉, 다중공선성이 높은 데이터를 제거하고 모델을 다시 만들어본결과 rmse 오차가 늘어나고 r_square값이 줄었으므로 변수를 제거하지 않고 그대로 가져가기로 한다.

p-value 값이 0.05를 넘어가면 다중공산성을 의심해볼만 함으로 p-value를 이용해 칼럼 제거 시도

In [22]:
import statsmodels.api as sm
result = sm.OLS(y_train,X_train).fit()
print(result.summary())

                                 OLS Regression Results                                
Dep. Variable:             total_sale   R-squared (uncentered):               82879.575
Model:                            OLS   Adj. R-squared (uncentered):          83036.342
Method:                 Least Squares   F-statistic:                             -528.7
Date:                Mon, 23 May 2022   Prob (F-statistic):                        1.00
Time:                        00:12:38   Log-Likelihood:                     -4.5884e+05
No. Observations:               19598   AIC:                                  9.177e+05
Df Residuals:                   19561   BIC:                                  9.180e+05
Df Model:                          37                                                  
Covariance Type:            nonrobust                                                  
                             coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------

확인해본결과 다중공산성 의심이 되는 p-value값이 0.05가 넘는 컬럼 총 9개('edu','edu_b','store_b','traffic_b','t_household_b','apt_66','apt_0.1b_under','apt_0.5b','apt_avg_price')를 지워서 validation으로 파악 해보기

In [23]:
X_p_train=X_train.drop(columns=['edu','edu_b','store_b','traffic_b','t_household_b','apt_66','apt_0.1b_under','apt_0.5b','apt_avg_price'])
X_p_val=X_val.drop(columns=['edu','edu_b','store_b','traffic_b','t_household_b','apt_66','apt_0.1b_under','apt_0.5b','apt_avg_price'])

In [25]:
lr = LinearRegression().fit(X_p_train, y_train)
y_p_train_hat = lr.predict(X_p_train)
print(y_train)
print(y_p_train_hat)
y_p_val_hat = lr.predict(X_p_val)
print(y_val)
print(y_p_val_hat)
print('performance for pTRAIN--------')
print('ptrain MAE : ', mean_absolute_error(y_train, y_p_train_hat))
print('ptrain RMSE : ', mean_squared_error(y_train, y_p_train_hat)**0.5)
print('ptrain R_square : ', r2_score(y_train, y_p_train_hat))

print('performance for pVAL--------')
print('val MAE : ', mean_absolute_error(y_val, y_p_val_hat))
print('val RMSE : ', mean_squared_error(y_val, y_p_val_hat)**0.5)
print('val R_square : ', r2_score(y_val, y_p_val_hat))


year
2021     6136905604
2021    21934115430
2021     9885562997
2021     2893691821
2021     4654682074
           ...     
2017     2528872295
2017      925316009
2017     2420562470
2017     8522048012
2017     1803437452
Name: total_sale, Length: 19598, dtype: int64
[7.81830180e+09 5.82791035e+09 2.00816468e+09 ... 5.47523150e+09
 7.12044931e+09 2.36600714e+09]
year
2021     5961650254
2021    17049532868
2021     9672498575
2021     2928283066
2021     4551502196
           ...     
2021     1876881076
2021      744236445
2021     1730866844
2021     7979637934
2021     1529373991
Name: total_sale, Length: 1088, dtype: int64
[7.81628992e+09 5.64499077e+09 1.92713579e+09 ... 6.70039836e+09
 5.92023283e+09 2.80959859e+09]
performance for pTRAIN--------
ptrain MAE :  2192293545.5999985
ptrain RMSE :  3558152061.963001
ptrain R_square :  0.5605839551250178
performance for pVAL--------
val MAE :  2280845288.454243
val RMSE :  3897053072.9069104
val R_square :  0.47075036399211334


모델 성능 rmse:3897053072.9069104, r_sqaure:0.47075036399211334

RMSE와 r_square를 비교해본결과 전보다 오차가 커짐을 알 수 있었다. 다중공선성이 의심되는 변수들을 제거해봤지만 성능은 좋아지지 않았다.

PCA를 이용하여 feature extraction을 해보겠다.(0.95를 맞춰주는 칼럼뽑기)

In [26]:
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA
scaler = StandardScaler()
scaler.fit(X_train)
X_train_scaled = scaler.transform(X_train)
X_val_scaled = scaler.transform(X_val)
pca = PCA(n_components= 0.95) 
pca.fit(X_train_scaled)
X_train_pca = pca.transform(X_train_scaled)
X_val_pca = pca.transform(X_val_scaled)
print("Original shape: {}".format(str(X_train_scaled.shape)))
print("Reduced shape: {}".format(str(X_train_pca.shape)))

Original shape: (19598, 38)
Reduced shape: (19598, 27)


38->27개로 총 11개의 feature가 줄어들었다.

In [27]:
lr = LinearRegression().fit(X_train_pca, y_train)
y_train_pca_hat = lr.predict(X_train_pca)
print(y_train)
print(y_p_train_hat)
y_val_pca_hat = lr.predict(X_val_pca)
print(y_val)
print(y_val_pca_hat)
print('performance for TRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_train_pca_hat))
print('train RMSE : ', mean_squared_error(y_train, y_train_pca_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_train_pca_hat))

print('performance for pcaVAL--------')
print('p MAE : ', mean_absolute_error(y_val, y_val_pca_hat))
print('p RMSE : ', mean_squared_error(y_val, y_val_pca_hat)**0.5)
print('p R_square : ', r2_score(y_val, y_val_pca_hat))

year
2021     6136905604
2021    21934115430
2021     9885562997
2021     2893691821
2021     4654682074
           ...     
2017     2528872295
2017      925316009
2017     2420562470
2017     8522048012
2017     1803437452
Name: total_sale, Length: 19598, dtype: int64
[7.81830180e+09 5.82791035e+09 2.00816468e+09 ... 5.47523150e+09
 7.12044931e+09 2.36600714e+09]
year
2021     5961650254
2021    17049532868
2021     9672498575
2021     2928283066
2021     4551502196
           ...     
2021     1876881076
2021      744236445
2021     1730866844
2021     7979637934
2021     1529373991
Name: total_sale, Length: 1088, dtype: int64
[7.64656395e+09 5.03954924e+09 1.89947182e+09 ... 7.23862611e+09
 5.48697753e+09 3.51663181e+09]
performance for TRAIN--------
train MAE :  2227377443.3105416
train RMSE :  3605188917.3883386
train R_square :  0.5488894780586748
performance for pcaVAL--------
p MAE :  2324603286.4577384
p RMSE :  3918378208.4358964
p R_square :  0.46494228283712646


모델성능 rmse:3918378208.4358964, r_square:0.46494228283712646

최종 모델 성능 확인하기 : 다중공산성을 줄이기 위해 vif 와 p-value를 이용하여 변수를 제거해보았지만 성능은 좋아지지 않았고, pca로 feature extraction을 시도해보았지만 이또한 성능은 좋아지지 않았다. 따라서 처음 데이터 그대로 linear regression을 적용한다.

train data로 모델을 만들어 test data로 성능 파악하기

In [28]:
lr = LinearRegression().fit(X_train, y_train)
y_train_hat = lr.predict(X_train)
print(y_train)
print(y_train_hat)
y_test_hat = lr.predict(X_test)
print(y_test)
print(y_test_hat)
print('performance for TRAIN--------')
print('train MAE : ', mean_absolute_error(y_train, y_train_hat))
print('train RMSE : ', mean_squared_error(y_train, y_train_hat)**0.5)
print('train R_square : ', r2_score(y_train, y_train_hat))

print('performance for test--------')
print('test MAE : ', mean_absolute_error(y_test, y_test_hat))
print('test RMSE : ', mean_squared_error(y_test, y_test_hat)**0.5)
print('test R_square : ', r2_score(y_test, y_test_hat))

year
2021     6136905604
2021    21934115430
2021     9885562997
2021     2893691821
2021     4654682074
           ...     
2017     2528872295
2017      925316009
2017     2420562470
2017     8522048012
2017     1803437452
Name: total_sale, Length: 19598, dtype: int64
[7.70896007e+09 5.53777310e+09 1.71273025e+09 ... 5.57186816e+09
 7.08910184e+09 2.37940539e+09]
year
2021     6164144274
2021    18215725779
2021     3130903256
2021     2770689276
2021     4831084227
           ...     
2021     2324569267
2021      619901202
2021     2596692568
2021     9209514193
2021     1703154580
Name: total_sale, Length: 1086, dtype: int64
[7.71219236e+09 5.36022902e+09 1.62466446e+09 ... 6.78242550e+09
 5.89541105e+09 2.76967663e+09]
performance for TRAIN--------
train MAE :  2191918759.671748
train RMSE :  3554464401.714925
train R_square :  0.5614943026057684
performance for test--------
test MAE :  2305539790.4288864
test RMSE :  3755264948.9428883
test R_square :  0.5111749733098415


기본 training data를 가지고 모델을 진행 한결과 test RMSE :  3755264948.9428883, test R_square :  0.5111749733098415 임을 알수있습니다.